plugins/wacom-usb/fu-wac-module-touch-id7.c

```diff
@@ -0,0 +1,362 @@
+ * with a start and end command each individual record.
+ * A single touch id7 firmware record can contain one or more blocks that have
+ * the raw data for writing.
+ */
+static gboolean
+writeSingleWTARecord(FuWacModule *self, struct WTAInfo *info, FuProgress *progress, GError **error) {
+	struct WTARecordHeader record_hdr;
+	guint8 command[11];
+
+	readWTARecordHeader(&record_hdr, info);
+
+	/* Start Firmware Record Command */
+	generateCommand(&record_hdr, FU_WAC_MODULE_COMMAND_START, info->op_id, command);
+	g_autoptr(GBytes) startSubBlob = g_bytes_new(command, sizeof(command));
+	if (!fu_wac_module_set_feature(self,
+					FU_WAC_MODULE_COMMAND_DATA,
+					startSubBlob,
+					fu_progress_get_child(progress),
+					FU_WAC_MODULE_ERASE_TIMEOUT,
+					error))
+		return FALSE;
+
+	info->op_id++;
+
+	/* Write Each Block */
+	for (guint32 i = 0; i < record_hdr.blockCount; i++) {
+		writeSingleWTABlock(self, info, progress, &record_hdr, error);
+	}
+
+	/* End Firmware Record Command */
+	generateCommand(&record_hdr, FU_WAC_MODULE_COMMAND_END, info->op_id, command);
+	g_autoptr(GBytes) endSubBlob = g_bytes_new(command, sizeof(command));
+	if (!fu_wac_module_set_feature(self,
+					FU_WAC_MODULE_COMMAND_DATA,
+					endSubBlob,
+					fu_progress_get_child(progress),
+					FU_WAC_MODULE_ERASE_TIMEOUT,
+					error))
+		return FALSE;
+
+	info->op_id++;
+
+	return TRUE;
+}
+
+/**
+ * fu_wac_module_touch_id7_write_firmware:
+ * 
+ * Start and End the overall update process for touch id7 firmware and the
+ * record(s) it contains.
+ * A touch id7 firmware will usually contian 3 firmware record(s) but could
+ * potentially have less or more.
+ */
+static gboolean
+fu_wac_module_touch_id7_write_firmware(FuDevice *device,
+				   FuFirmware *firmware,
+				   FuProgress *progress,
+				   FwupdInstallFlags flags,
+				   GError **error)
+{
+	FuWacModule *self = FU_WAC_MODULE(device);
+	g_autoptr(GBytes) fwRead = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+	struct WTAInfo info;
+	struct WTAFileHeader file_hdr;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 2, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 97, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1, NULL);
+
+	g_debug("using element at addr 0x%0x", (guint)fu_firmware_get_addr(firmware));
+
+	fwRead = fu_firmware_get_bytes(firmware, error);
+	if (fwRead == NULL)
+		return FALSE;
+
+	/* start, which will erase the module */
+	if (!fu_wac_module_set_feature(self,
+				       FU_WAC_MODULE_COMMAND_START,
+				       NULL,
+				       fu_progress_get_child(progress),
+				       FU_WAC_MODULE_ERASE_TIMEOUT,
+				       error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* Set basic info */
+	info.data = g_bytes_get_data(fwRead, &info.len);
+	info.op_id = 1;
+
+	readWTAFileHeader(&file_hdr, &info);
```

Comment:
this really isn't in the GLib style. Please return boolean, and accept a GError as the last parameter. You can pass `WTAFileHeader` by ref. I also wonder if this makes sense to split off this as a `FuWacModuleTouchId7Firmware` object -- and then we can unit test parsing it.

Reply:
For the read functions would it warrant returning a boolean or passing GError? The most they do is move the data object forward or call fu_memread_uintXX methods. There's not much to catch in terms of failures from what I can think of.

Reply:
I think moving it to subclass FuFirmware will allow the` ->parse()` vfunc to do a lot of the heavy lifting.

Reply:
Is there a good example of a FuFirmware subclass I can look at to better understand it and how it's best implemented?

Reply:
I think any of the existing FuFirmware subclasses that parse binary data would be fine -- e.g. FuFmapFirmware or FuFdtFirmware. There are a ton of developer docs here: https://fwupd.github.io/libfwupdplugin/tutorial.html#creating-a-new-firmware-type and https://fwupd.github.io/libfwupdplugin/class.Firmware.html

Reply:
@Joshua-Dickens would you like me to add you some template code?

Reply:
Quick before that, looking through what you sent parse is meant to go over the entire file first, then write takes care of the various images gathered by the parse. 

But the way we settled on writing this firmware code to be the simplest/most readable was to parse as we go such as.
(File Header Parse) -> Individual Firmware -> (Firmware Header Parse) -> Individual Blocks -> (Block Header Parse) -> Write

We initially were parsing it all at once then writing but because the new file type we're handling (wtg) can have any number of firmware and any number of blocks in each firmware along with the commands being sent to the tablet relying on information from the firmware header and the block header + data, we found that parsing alongside the writing of each block made for cleaner overall code.

I can work it back to parsing it all at once and then writing it, but I first want to ask if there is a way to keep the process of how we're doing it right now while fitting it into your suggested changes.

Reply:
What we normally do is parse the entire file (to verify it's complete and compatible) and then write the prepared object to the device. If you're asking the device for the next block that's harder, as we still need to do an accurate percentage-complete signal.

Reply:
With the new firmware it's a bit harder to verify things are correct with how the new firmware type (.wtg) is structured.

How the file is structured means we only get that information as we go, so when we read the file header it tells us how many firmware images we have but not how many blocks we have, it's not until we get to the header of the individual firmware that we learn how many blocks are in that specific firmware image, and it isn't until we go to the individual block header that we get the size for the block that we have no real way of verifying since there's no distinct symbol or mark of an end for the block and the start of a new block or firmware. And since there's no distinct symbol/data to identify the firmware images/blocks there's no good way to check if we have the proper number of firmware/blocks in the file in the first place. 

The only static symbol in the file that will always be constant is a string that reads "WTA" at the very beginning of the file. Everything else is values we have no static expected values for meaning we have to trust the data we get during parse.

For verification that the data is correct we can only really rely on the start and end commands we're sending to the tablet for each individual firmware image with which the tablet verifies if the firmware image wrote correctly and give a response if it did, or does not respond and cause a timeout which has fwupd abort the update properly and the tablet will boot into a safe mode which disables touch until a clean touch firmware install happens but otherwise doesn't affect functionality of the tablet.

Originally we wanted to parse everything at first, but since there wasn't any real way to truly verify the data is good/complete it made less sense to us to keep the process separate and instead to make the code simpler and more readable we would write the blocks as we got to them and let the tablet verify using the additional start and end commands for each firmware image. That way if for some reason the data within the block got corrupted or if the data in the header was wrong, the tablet would catch it and abort safely.

If you want to preserve the styling of parsing everything first then writing I absolutely can swap it over to that style. But I just wanted to quick give you an overview of why we swapped away from that and the difficulties in verifying the data so that you have the full context of why we made the decisions we did in how we structured the code.

Reply:
Okay -- it's your plugin and you can do what you want. But -- I still think a custom FuFirmware subclass makes sense as then you can pre-parse the format and get everything in easy-to-push packets of data for the actual write -- and you can also test and fuzz the parser easily as we need to be careful of malicious data being fed to fwupd.